server/go/websocket/router.go

```diff
@@ -0,0 +1,47 @@
+package websocket
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+)
+
+type MessageExecutor func(*websocket.Conn, Message)
+
+type Router struct {
+	routes map[string]MessageExecutor
+}
+
+func NewRouter() *Router {
+	return &Router{
+		routes: make(map[string]MessageExecutor),
+	}
+}
+
+func (r *Router) Add(messageType string, executor MessageExecutor) {
+	r.routes[messageType] = executor
+}
+
+func (r *Router) ListenAndServe(addr string) {
```

Comment:
noob question, where does r come from here? Is it somehow getting injected?

Reply:
The function signature means that the function is attached to the struct `Router`. That's the go way to create something similar to methods. You can think of the `r` as `this` (though not quite the same).  You can read [a more precise definition](https://go.dev/tour/methods/1)